### PR TITLE
The great artifacts multi registries overhaul.

### PR DESCRIPTION
**Pull request by BillyONeal**
User facing intentional behavior changes:
* The --registry switch has been removed. There was basically nothing we could do with a registry for which we had no name.
* Commands which speak registries / IDs now combine the set of names of registries from the global set and the local project set.
* Console output now always displays registries with the following rules:
1. If the registry URI is in the project, the name from the project is used.
2. Otherwise, if the registry URI is is not in the project, and the name of the registry in the global configuration does not conflict with the project, the global configuration's name is used.
3. Otherwise, the URI enclosed in []s is used. This is a signal to the user that they will need to take some form of manual action.
* The command x-regenerate now requires a local file path.
* Several places that used to break when handed relative paths, like x-regenerate, now tolerate this situation correctly.
* If an artifact declares a dependency with explicit registry, and that registry's name is not declared in that artifact, resolution fails. (We no longer will try to find that name in the project / global configuration)
* Commands that need activation options like `add`, and `remove`, no longer try to activate the project.
* Artifacts are now installed and otherwise enumerated in topological sorted order. This is the other half of https://devdiv.visualstudio.com/DevDiv/_boards/board/t/Vcpkg%20Package%20Manager/Stories/?workitem=1510278
* There is no longer a built in registry named 'default'.
* Except for index.yaml, a few remaining yaml hangers-on have been replaced with JSON. This may invalidate installed artifacts in the wild. This is necessary for the medium-term goal of deleting the YAML implementation from the product entirely.
* index.yaml no longer begins with 'THIS_IS_NOT_A_MANIFEST_ITS_AN_INDEX_STRING'
* A bunch of newlines in output are now gone. I can see an argument to add some of these back but we certainly weren't principled about it before and it caused varying numbers of newlines to come out.

Bug fixes:
* The console output no longer gets confused and displays [URL]'d registry names (except as denoted by case (3) above). This was https://devdiv.visualstudio.com/DevDiv/_boards/board/t/Vcpkg%20Package%20Manager/Stories/?workitem=1494994

TODO before un-drafting this review:
* <s>Fix recording of the 'requested' version in `add`</s>
* <s>Audit for FIXMEs</s>
* <s>Audit for local variable names like 'artifacts' that should now be 'resolutions' or similar.</s>
* <s>See if the 'initialSelections' machinery in resolveDependencies can be made less painful.</s> I wanted to do this but at this point this work is blocking people so I want to merge what works now. Will probably clean that up a lot in a subsequent review.
* <s>Separate parts of this into smaller blocks that can be more easily reviewed where that makes sense.</s>
  * <s>https://github.com/microsoft/vcpkg-tool/pull/692</s>
  * <s>https://github.com/microsoft/vcpkg-tool/pull/693</s>
* Write up better 'line by line' ish commit notes to make it easier to review.
* <s>Tests</s>

### Fun demonstrations

Demonstrating registry naming edge cases:
Registry naming edge cases demonstration

Demonstrating correct registry name and correct dependency star:
Registry name and dependency star demonstration

Demonstrating topological sort. Note that 'headers' is a diamond dependency:

Topological sort demonstration

### CODE REVIEW

**Comment by BillyONeal** [comment]
Known issue that will not be fixed in this change: version resolution is broken:
image

File: ce/ce/artifacts/artifact.ts [hunk 1]
```diff
@@ -3,113 +3,96 @@
...
+  throw new Error(i`Invalid artifact id '${id}'`);
+}
 
-    // load the registries from the project file
-    for (const [name, registry] of this.metadata.registries) {
-      const reg = await session.loadRegistry(registry.location.get(0), registry.registryKind || 'artifact');
-      if (reg) {
-        this.registries.add(reg, name);
-      }
+export async function buildRegistryResolver(session: Session, registries: RegistriesDeclaration) {
```
Comment (BillyONeal) [inline ce/ce/artifacts/artifact.ts:35]: Should the session function die?

File: ce/ce/registries/registries.ts [hunk 2]
```diff
@@ -2,126 +2,238 @@
...
+  getRegistryByUri(registryUri: string) {
+    return this.#uriToRegistry.get(registryUri);
   }
 
-  getRegistryName(registry: Registry): string {
-    for (const [name, reg] of this.registries) {
-      if (reg === registry && name.indexOf('://') === -1) {
-        return name;
-      }
+  entries() { return this.#uriToRegistry.values(); }
```
Comment (ras0219-msft) [inline ce/ce/registries/registries.ts:68]: I can't find any uses of this function.

File: ce/ce/registries/registries.ts [hunk 3]
```diff
@@ -2,126 +2,238 @@
...
+
+  has(registryUri: string) { return this.#uriToRegistry.has(registryUri); }
+
+  add(uri: Uri, registry: Registry) {
```
Comment (ras0219-msft) [inline ce/ce/registries/registries.ts:72]:
```suggestion
  // Exposed for testing
  add(uri: Uri, registry: Registry) {
```

File: ce/ce/registries/registries.ts [hunk 4]
```diff
@@ -2,126 +2,238 @@
...
+  readonly #knownUris: Set<string>;
+  readonly #uriToName: Map<string, string>;
+  readonly #nameToUri: Map<string, string>;
+
+  private addMapping(name: string, uri: string) {
+    this.#uriToName.set(uri, name);
+    this.#nameToUri.set(name, uri);
+  }
+
+  constructor(parent: RegistryDatabase | RegistryResolver) {
```
Comment (ras0219-msft) [inline ce/ce/registries/registries.ts:125]: It seems unfortunate that the constructor needs to be complicated just to support the "clone" operation in `combineWith`.

File: ce/ce/session.ts [hunk 5]
```diff
@@ -125,139 +118,31 @@ export class Session {
...
+      return this.fileSystem.parse(location);
     }
 
-    return false;
-  }
-
-  async isRemoteRegistry(location: Uri | string): Promise<boolean> {
-    return this.parseUri(location).scheme === 'https';
+    // Relative path
+    return this.currentDirectory.join(location);
```
Comment (ras0219-msft) [inline ce/ce/session.ts:135]: Is this making URIs relative to the file they're found in, the top level file, or the current working directory?

File: ce/ce/session.ts [hunk 6]
```diff
@@ -329,39 +214,30 @@ export class Session {
   async deactivate() {
     const previous = this.environment[undo];
     if (previous && this.postscriptFile) {
-      const deactivationDataFile = this.parseUri(previous);
+      const deactivationDataFile = this.fileSystem.file(previous);
       if (deactivationDataFile.scheme === 'file' && await deactivationDataFile.exists()) {
```
Comment (ras0219-msft) [inline ce/ce/session.ts:218]:
```suggestion
      if (await deactivationDataFile.exists()) {
```
This is also the usual exists+read antipattern

File: ce/ce/amf/metadata-file.ts [hunk 7]
```diff
@@ -17,37 +14,31 @@ import { Contacts } from './contact';
...
-  }
 
-  async init(session: Session): Promise<MetadataFile> {
-    await this.demandBlock.init(session);
-    return this;
   }
 
-  static async parseMetadata(uri: Uri, session: Session, registry?: Registry): Promise<MetadataFile> {
-    return MetadataFile.parseConfiguration(uri.path, await uri.readUTF8(), session, registry);
+  static async parseMetadata(filename: string, uri: Uri, session: Session, registryUri?: Uri): Promise<MetadataFile> {
```
Comment (ras0219-msft) [inline ce/ce/amf/metadata-file.ts:25]: Every call site for this function looks like:
```ts
MetadataFile.parseMetadata(x.fsPath, x, ...)
```
This argues to me that this convenience overload should just take the `uri` -- if a caller really wants to control the filename, they can `await uri.readUTF8()` and call the lower level one.

Secondly, `fsPath` is explicitly documented as not for display purposes:

>      * Returns a string representing the corresponding file system path of this URI.
     * Will handle UNC paths, normalizes windows drive letters to lower-case, and uses the
     * platform specific path separator.
     *
     * * Will *not* validate the path for invalid characters and semantics.
     * * Will *not* look at the scheme of this URI.
     * * The result shall *not* be used for display purposes but for accessing a file on disk.

Maybe we don't care and want to use it anyway, but it should be considered.

**Review by ras0219-msft — APPROVED** [review]

File: ce/ce/artifacts/artifact.ts [hunk 8]
```diff
@@ -3,113 +3,96 @@
...
+  return result;
+}
 
-        if (recurse) {
-          // process it's dependencies too.
-          await artifact.resolveDependencies(artifacts);
-        }
-      }
-    }
+export abstract class ArtifactBase {
```
[the remark this answers is not in the document]
Comment (BillyONeal) [reply ce/ce/artifacts/artifact.ts:59]: Yeah, I don't want to go there for purposes of this PR for now.

File: ce/ce/artifacts/activation.ts [hunk 9]
```diff
@@ -548,14 +548,14 @@ export class Activation {
...
+  async activate(artifacts: Iterable<ResolvedArtifact>, currentEnvironment: Record<string, string | undefined>, shellScriptFile: Uri | undefined, undoEnvironmentFile: Uri | undefined, msbuildFile: Uri | undefined, json: Uri | undefined) {
     let undoDeactivation = '';
     const scriptKind = extname(shellScriptFile?.fsPath || '');
 
     // load previous activation undo data
     const previous = currentEnvironment[undoVariableName];
     if (previous && undoEnvironmentFile) {
-      const deactivationDataFile = this.#session.parseUri(previous);
+      const deactivationDataFile = this.#session.fileSystem.file(previous);
       if (deactivationDataFile.scheme === 'file' && await deactivationDataFile.exists()) {
```
[the remark this answers is not in the document]
Comment (BillyONeal) [reply ce/ce/artifacts/activation.ts:559]: Also nuked bad exists call.

File: ce/ce/artifacts/artifact.ts [hunk 10]
```diff
@@ -292,6 +243,138 @@ export class ProjectManifest extends ArtifactBase {
...
+
+export interface ResolvedArtifact {
+  artifact: ArtifactBase,
+  uniqueId: string,
+  initialSelection: boolean,
+  depth: number,
+  priority: number
+}
+
+export async function resolveRegistries(registryResolver: RegistryResolver, initialParents: Array<ArtifactBase>): Promise<Array<Registry | undefined>> {
```
[the remark this answers is not in the document]
Comment (BillyONeal) [reply ce/ce/artifacts/artifact.ts:258]: Because my brain initially thought this was going to call loadRegistry 😅. Fixed!

File: ce/ce/artifacts/artifact.ts [hunk 11]
```diff
@@ -292,6 +243,138 @@ export class ProjectManifest extends ArtifactBase {
...
+    result.push(registry);
+  }
+
+  return result;
+}
+
+export async function resolveDependencies(session: Session, registryResolver: RegistryResolver, initialParents: Array<ArtifactBase>, dependencyDepth: number): Promise<Array<ResolvedArtifact>> {
+  let depth = 0;
+  let nextDepthParents: Array<Registry | undefined> = await resolveRegistries(registryResolver, initialParents);
+  let currentParents: Array<Registry | undefined> = [];
```
[the remark this answers is not in the document]
Comment (BillyONeal) [reply ce/ce/artifacts/artifact.ts:276]: Done!

File: ce/ce/artifacts/artifact.ts [hunk 12]
```diff
@@ -292,6 +243,138 @@ export class ProjectManifest extends ArtifactBase {
...
+
+    result.push(registry);
+  }
+
+  return result;
+}
+
+export async function resolveDependencies(session: Session, registryResolver: RegistryResolver, initialParents: Array<ArtifactBase>, dependencyDepth: number): Promise<Array<ResolvedArtifact>> {
+  let depth = 0;
+  let nextDepthParents: Array<Registry | undefined> = await resolveRegistries(registryResolver, initialParents);
```
[the remark this answers is not in the document]
Comment (BillyONeal) [reply ce/ce/artifacts/artifact.ts:275]: Ah, `Select` is called `map` here. Thanks!

File: ce/ce/artifacts/artifact.ts [hunk 13]
```diff
@@ -292,6 +243,138 @@ export class ProjectManifest extends ArtifactBase {
...
+  const results = new Array<ResolvedArtifact>();
+  for (const [uniqueId, artifact] of resultSet) {
+    const order = orderer.get(uniqueId);
+    if (order) {
+      results.push({
+        'artifact': artifact,
+        'uniqueId': uniqueId,
+        'initialSelection': initialSelections.has(uniqueId),
+        'depth': order[0],
+        'priority': order[1]
```
[the remark this answers is not in the document]
Comment (BillyONeal) [reply ce/ce/artifacts/artifact.ts:364]: Done!

File: ce/ce/artifacts/artifact.ts [hunk 14]
```diff
@@ -292,6 +243,138 @@ export class ProjectManifest extends ArtifactBase {
...
+        let dependencyRegistry: Registry;
+        if (dependencyRegistryDeclaredName) {
+          const maybeRegistry = (await subject.buildRegistryResolver()).getRegistryByName(dependencyRegistryDeclaredName);
+          if (!maybeRegistry) {
+            throw new Error(i`While resolving dependencies of ${subjectId}, ${dependencyRegistryDeclaredName} in ${idOrShortName} could not be resolved to a registry.`);
+          }
+
+          dependencyRegistry = maybeRegistry;
+        } else {
+          if (!subjectParentRegistry) {
```
[the remark this answers is not in the document]
Comment (BillyONeal) [reply ce/ce/artifacts/artifact.ts:328]: Yeah :/

File: ce/ce/artifacts/artifact.ts [hunk 15]
```diff
@@ -292,6 +243,138 @@ export class ProjectManifest extends ArtifactBase {
...
+        // already visited
+        continue;
+      }
+
+      resultSet.set(subjectUniqueId, subject);
+      for (const [idOrShortName, version] of linq.entries(subject.applicableDemands.requires)) {
+        const [dependencyRegistryDeclaredName, dependencyId] = parseArtifactDependency(idOrShortName);
+        let dependencyRegistry: Registry;
+        if (dependencyRegistryDeclaredName) {
+          const maybeRegistry = (await subject.buildRegistryResolver()).getRegistryByName(dependencyRegistryDeclaredName);
```
[the remark this answers is not in the document]
Comment (BillyONeal) [reply ce/ce/artifacts/artifact.ts:321]: Agreed, fixed!

File: ce/ce/session.ts [hunk 16]
```diff
@@ -125,139 +118,31 @@ export class Session {
...
   }
 
-  parseName(id: string): [string, string] {
-    const parts = id.split(':');
-    switch (parts.length) {
-      case 0:
-        throw new Error(i`Invalid artifact id '${id}'`);
-      case 1:
-        return ['default', id];
+  async loadDefaultRegistryResolver(manifest: ProjectManifest | undefined) : Promise<RegistryResolver> {
```
[the remark this answers is not in the document]
Comment (BillyONeal) [reply ce/ce/session.ts:138]: The problem is that makes `undefined` checks in callers painful since ProjectManifest is one of the things likely to be `undefined`. For example:

```diff
diff --git a/ce/ce/cli/commands/acquire.ts b/ce/ce/cli/commands/acquire.ts
index 23253dfb..95a227ed 100644
--- a/ce/ce/cli/commands/acquire.ts
+++ b/ce/ce/cli/commands/acquire.ts
@@ -44,7 +44,8 @@ export class AcquireCommand extends Command {
       return false;
     }

-    const resolver = await session.loadDefaultRegistryResolver(await this.project.manifest);
+    const project = await this.project.manifest;
+    const resolver = (await project?.buildRegistryResolver())?.combineWith(session.globalRegistryResolver) ?? session.globalRegistryResolver;
     const resolved = await selectArtifacts(session, new Map(this.inputs.map((v, i) => [v, versions[i] || '*'])), resolver, 1);        
     if (!resolved) {
       debug('No artifacts selected - stopping');
```

[the remark this answers is not in the document]
Comment (BillyONeal) [reply ce/ce/session.ts:138]: I suppose I could flip combineWith's sense since session.globalRegistryResolver won't be undefined...

File: ce/ce/cli/commands/add.ts [hunk 17]
```diff
@@ -58,30 +52,45 @@ export class AddCommand extends Command {
...
+          // the registry isn't known yet to the project, try to declare it
+          registryName = session.globalRegistryResolver.getRegistryName(registryUri);
+          if (!registryName) {
+            throw new Error(i`Tried to add an artifact [${registryUri.toString()}]:${artifact.id} but could not determine the registry to use.`);
+          }
+
+          const conflictingRegistry = projectRegistries.getRegistryByName(registryName);
+          if (conflictingRegistry) {
+            throw new Error(i`Tried to add registry ${registryName} as ${registryUri.toString()}, but it was already ${conflictingRegistry.location.toString()}. Please add ${registryUri.toString()} to this project manually and reattempt.`);
+          }
```
[the remark this answers is not in the document]
Comment (BillyONeal) [reply ce/ce/cli/commands/add.ts:79]: I'm not sure what you mean. Combined resolvers don't have historical information about from which parent a registry comes.

File: ce/ce/cli/commands/add.ts [hunk 18]
```diff
@@ -58,30 +52,45 @@ export class AddCommand extends Command {
...
+
+          projectManifest.metadata.registries.add(registryName, artifact.registryUri, 'artifact');
```
[the remark this answers is not in the document]
Comment (BillyONeal) [reply ce/ce/cli/commands/add.ts:81]: No, it was missing an additional add. Fixed!

File: ce/ce/cli/commands/add.ts [hunk 19]
```diff
@@ -58,30 +52,45 @@ export class AddCommand extends Command {
...
-      const fulfilled = artifact.version.toString();
-      const v = requested !== fulfilled ? `${requested} ${fulfilled}` : fulfilled;
-      projectManifest.metadata.requires.set(artifact.reference, <any>v);
     }
 
     // write the file out.
     await projectManifest.metadata.save();
-
-    return await activateProject(projectManifest, this.commandLine);
+    return true;
```
[the remark this answers is not in the document]
Comment (BillyONeal) [reply ce/ce/cli/commands/add.ts:94]: I think we would need to track activation settings to do that correctly.

File: ce/ce/cli/commands/find.ts [hunk 20]
```diff
@@ -54,16 +46,16 @@ export class FindCommand extends Command {
       })) {
         if (!this.version.isRangeOfVersions) {
           // if the user didn't specify a range, just show the latest version that was returned
-          artifacts = [artifacts[0]];
+          artifactVersions = [artifactVersions[0]];
```
[the remark this answers is not in the document]
Comment (BillyONeal) [reply ce/ce/cli/commands/find.ts:49]: I believe that is a clearly established postcondition but I applied this change anyway.

File: ce/ce/cli/format.ts [hunk 21]
```diff
@@ -8,27 +8,31 @@ export function projectFile(uri: Uri): string {
...
 
-export function artifactIdentity(registryName: string, identity: string, alias?: string) {
-  if (alias) {
-    return `${whiteBright(registryName)}:${yellow.dim(identity.substr(0, identity.length - alias.length))}${yellowBright(alias)}`;
+export function prettyRegistryName(registryName: string) {
+  return `${whiteBright(registryName)}`;
+}
+
+export function artifactIdentity(registryName: string, identity: string, shortName: string) : string {
+  if (shortName) {
```
[the remark this answers is not in the document]
Comment (BillyONeal) [reply ce/ce/cli/format.ts:16]: Nope! Removed.

File: ce/ce/cli/commands/version.ts [hunk 22]
```diff
@@ -37,7 +37,7 @@ export class VersionCommand extends Command {
   check = new Check(this);
   update = new Update(this);
 
-  versionUrl = session.parseUri('https://aka.ms/vcpkg-ce.version');
+  versionUrl = session.fileSystem.parse('https://aka.ms/vcpkg-ce.version');
```
[the remark this answers is not in the document]
Comment (BillyONeal) [reply ce/ce/cli/commands/version.ts:40]: Renamed!

File: ce/ce/registries/ArtifactRegistry.ts [hunk 23]
```diff
@@ -169,6 +160,6 @@ export abstract class ArtifactRegistry implements Registry {
   }
 
   async save(): Promise<void> {
-    await this.indexYaml.writeFile(Buffer.from(`${THIS_IS_NOT_A_MANIFEST_ITS_AN_INDEX_STRING}\n${serialize(this.index.serialize()).replace(/\s*(\d*,)\n/g, '$1')}`));
+    await this.indexYaml.writeFile(Buffer.from(serialize(this.index.serialize()).replace(/\s*(\d*,)\n/g, '$1')));
```
[the remark this answers is not in the document]
Comment (BillyONeal) [reply ce/ce/registries/ArtifactRegistry.ts:163]: Done.

File: ce/ce/registries/registries.ts [hunk 24]
```diff
@@ -2,126 +2,238 @@
...
-import { linq } from '../util/linq';
 import { Uri } from '../util/uri';
+import { LocalRegistry } from './LocalRegistry';
+import { RemoteRegistry } from './RemoteRegistry';
 
-export class Registries implements Iterable<[Registry, Array<string>]> {
-  protected registries: Map<string, Registry> = new Map();
+export interface SearchCriteria {
+  idOrShortName?: string;
+  version?: string
```
[the remark this answers is not in the document]
Comment (BillyONeal) [reply ce/ce/registries/registries.ts:15]: Done!

File: ce/ce/registries/registries.ts (hunk 2 above)
Comment (BillyONeal) [reply ce/ce/registries/registries.ts:68]: Deleted.

File: ce/ce/registries/registries.ts (hunk 4 above)
Comment (BillyONeal) [reply ce/ce/registries/registries.ts:125]: idk, I think the clone operation makes sense to offer in general, and attempting to hoist it out either forces removing `readonly` from the set and maps, or giving up being able to call the ctor to clone those.

File: ce/ce/session.ts (hunk 5 above)
Comment (BillyONeal) [reply ce/ce/session.ts:135]: Current working directory. Hmmmm.... maybe we shouldn't have that fallback and fix console arguments to be parsed with something else.

File: ce/ce/amf/metadata-file.ts (hunk 7 above)
Comment (BillyONeal) [reply ce/ce/amf/metadata-file.ts:25]: This change is just getting that 'context' junk out of display operations and hoisting that problem out of this and into the immediate callers. I agree, the unwrapping into fsPath is a serious problem. I would like to see the over-reliance on URIs for things that aren't URIs burninated, but that would be its own very large but also very mechanical change so I don't want to combine it with this.

File: ce/ce/session.ts (hunk 5 above)
Comment (BillyONeal) [reply ce/ce/session.ts:135]: OK, removed CWD fallback here, audited all callers and verified that they never expected relative paths, and also audited all callers of `filesystem.file()` and added Path.resolve calls to them where reasonable.

File: ce/ce/amf/metadata-file.ts (hunk 7 above)
Comment (BillyONeal) [reply ce/ce/amf/metadata-file.ts:26]: Is https://github.com/microsoft/vcpkg-tool/pull/690/commits/dcff983adf751b29de88335015e36eb2f7722061 acceptable?

Comment (ras0219-msft) [reply ce/ce/amf/metadata-file.ts:26]: Seems like a fine change to me

File: ce/ce/cli/commands/add.ts (hunk 17 above)
[the remark this answers is not in the document]
Comment (ras0219-msft) [reply ce/ce/cli/commands/add.ts]: Today they do not, but it seems to me like that would be a good refactor in the future (de-generalize the different resolvers).

File: ce/ce/cli/commands/add.ts (hunk 19 above)
[the remark this answers is not in the document]
Comment (ras0219-msft) [reply ce/ce/cli/commands/add.ts:95]: Even without activation settings, it would still be good to tell the user that they need to reactivate -- but point taken that we can't give them the correct `vcpkg activate`.
